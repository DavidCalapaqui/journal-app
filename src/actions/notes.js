import { db } from "../firebase/firebase-config";
import { collection, addDoc, updateDoc, doc, deleteDoc } from "firebase/firestore";
import {types} from '../types/types';

import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";



export const startNewNote = () =>{
    
    
    return async (dispatch, getState) =>{
        const {uid} = getState().auth;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const doc = await addDoc(collection(db, `${uid}/journal/notes`), newNote);
            //console.log('DOC REF: ', doc);

            dispatch(activeNote(doc.id, newNote));
            dispatch(addNewNote(doc.id, newNote));
        } catch (error){
            console.error('ERROR AL INTENTAR GRABAR UNA COLECCION: ',error);
        } 
       
    }
}

export const startLoadingNotes = (uid) =>{
    return async(dispatch)=>{
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}

export const activeNote = (id, note) =>({
    type: types.notesActive,
    payload:{
        id, 
        ...note
    }
})

export const addNewNote = (id, note)=>({
    type: types.notesAddNew,
    payload:{
        id,
        ...note
    }
})

export const setNotes = (notes) =>({
    type: types.notesLoad,
    payload: notes
})


export const startSaveNote = (note) =>{
    return async (dispatch, getState)=>{
        const {uid} = getState().auth;

        if(!note.url){
            delete note.url;
        }

        const noteToFirestore = {...note};
        delete noteToFirestore.id;
        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
        try {
             await updateDoc(noteRef, noteToFirestore);
        } catch (error) {
            console.error('ERROR AL ACTUALIZAR NOTA: ',error);
        }

        dispatch(refreshNote(note.id, note));
        Swal.fire('Saved', note.title, 'success')
    }
}

export const refreshNote = (id, note) =>({
    type: types.notesUpdated,
    payload:{
        id, 
        note:{
            id,
            ...note
        }
    }
})


export const startUploading = (file) =>{

    return async (dispatch,getState ) =>{
        const {active:activeNote} = getState().notes;
        Swal.fire({
            title: 'Uploading',
            text: 'Please wait...',
            allowOutsideClick: false,
            onBeforeOpen: ()=>{ 
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote))
        Swal.close();
    }
}


export const startDeleting = (id) =>{
    return async (dispatch, getState)=>{
        const uid = getState().auth.uid;

        const noteRef = doc(db, `${uid}/journal/notes/${id}`);

        try {
            await deleteDoc(noteRef);
            dispatch(deleteNote(id));
            Swal.fire('Note removed', 'The note has been removed on database', 'success')

        } catch (error) {
            Swal.fire('Error', error.message, 'error')
        }

    }
}

export const deleteNote = (id) =>({
    type: types.notesDelete,
    payload: id
})

//limpiar las notas del usuario en sesión cuando cierre sesión
export const notesLogout =()=>({
    
    type: types.notesLogoutCleaning
})