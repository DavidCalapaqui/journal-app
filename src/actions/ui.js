import { types } from "../types/types";

//ACCIONES PARA ERRORES EN CAMPOS
export const setError = (err) =>({
    type: types.uiSetError,
    payload: err
})

export const removeError = () =>({
    type: types.uiRemoveError,
})

//ACCIONES PARA CARGANDO
export const startLoading = () =>({
    type: types.uiStartLoading,
})

export const finishLoading = () =>({
    type: types.uiFinishLoading,
})