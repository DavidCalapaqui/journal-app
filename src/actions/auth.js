
import {getAuth, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { finishLoading, startLoading } from './ui';
import { notesLogout } from './notes';

//accion
export const startLoginEmailPassword = (email, password) =>{
    //retorna un callback 
    return (dispatch)=>{
        //dispatch startLogin
        dispatch(startLoading());

        const auth = getAuth();
        return signInWithEmailAndPassword(auth, email, password)
        .then(async ({user})=>{
            dispatch(login(user.uid, user.displayName));
            
            dispatch(finishLoading());
            //dispatch stopLoggin
            //console.log('EMAIL & PASS - USER SIGIN: \n',user.displayName, user.uid);
        }).catch(e=>{
            Swal.fire('Error', e.message, 'error' );
            }
        );
    }

}

export const startRegisterWithEmailPasswordName = (email, password, name) =>{
    return(dispatch)=>{
        const auth = getAuth();
        createUserWithEmailAndPassword(auth, email, password)
        .then(async({user})=>{ 
            //dispatch(login(user.uid, user.displayName));
            await updateProfile(user, {displayName: name});
            dispatch(user.uid, user.displayName);
            console.log(user);
        })
        .catch(e=>{
            Swal.fire('Error', e.message, 'error' );
        })
    }

}

//accion
export const startGoogleLogin = () =>{
    return (dispatch)=>{
       const auth = getAuth();
       signInWithPopup(auth, googleAuthProvider)
       .then(({user})=>{
            console.log('GOOGLE - USER SIGIN: \n', user.displayName, user.uid);
           dispatch(login(user.uid, user.displayName));
       })
    }
}
//accion
export const login = (uid, displayName) =>{
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

//acción asíncrona a firebase
export const startLogout = () =>{
    const auth = getAuth();

    return async (dispatch)=>{
        await auth.signOut();
        dispatch(logout()); 
        //disparar accion para limpiar notas
        dispatch(notesLogout());
    }
}

export const logout = () =>({
    type: types.logout
});