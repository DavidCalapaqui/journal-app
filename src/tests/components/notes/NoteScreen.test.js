import React from 'react';
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';

import { activeNote } from '../../../actions/notes';
import { NoteScreen } from '../../../components/notes/NoteScreen';
//import {MemoryRouter} from 'react-router-dom'; //finjir rutas



const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock('../../../actions/notes', ()=>({
    activeNote: jest.fn(),
}))


const initState = {
    auth: {
        uid: 'GcuFhm6rFWeVQm9bfJpFQoqiQ9C2',
        name: 'David'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes:{
        notes: [
            {
                id    :"ENqiu5vFclsHuA5geru3",
                title :"Note title!!!",
                date  :1644273441664,
                body  :"This is anther note!!!",
                url   :"https://res.cloudinary.com/api-node-cloud/image/upload/v1644363562/journal-photos/r4rn9t0sczxtzhrbewfl.png"
            }
        ],
        active: {
            id    :"ENqiu5vFclsHuA5geru3",
            title :"Note title!!!",
            date  :1644273441664,
            body  :"This is anther note!!!",
            url   :"https://res.cloudinary.com/api-node-cloud/image/upload/v1644363562/journal-photos/r4rn9t0sczxtzhrbewfl.png"
        }
    }
}

//simulacion del store
let store = mockStore(initState);
//reemplazar la funcionalidad del store por una funcion de jest para poder evaluarla
store.dispatch = jest.fn();

const wrapper = mount(
    
    <Provider store={store}>
            <NoteScreen/>
    </Provider>
        
);



describe('Pruebas en NoteScreen', ()=>{

    test('Debe de mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();

    })

    //modificar caja de texto para activar nota

    test('Debe disparar el activeNote', () => { 
        wrapper.find('input[name="title"]').simulate('change',{
            terget: {
                name: 'title',
                value: 'hola otra vez'
            }
        })    

        expect(activeNote).toHaveBeenLastCalledWith(
            "ENqiu5vFclsHuA5geru3",
            {
                id    :"ENqiu5vFclsHuA5geru3",
                title :"Note title!!!",
                date  :1644273441664,
                body  :"This is anther note!!!",
                url   :"https://res.cloudinary.com/api-node-cloud/image/upload/v1644363562/journal-photos/r4rn9t0sczxtzhrbewfl.png"
            }
        );
    })


})