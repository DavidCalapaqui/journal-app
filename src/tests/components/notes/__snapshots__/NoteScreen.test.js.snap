// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en NoteScreen Debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content animate__animated animate__bounceIn"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de enero 2022
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Note title!!!"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="This is anther note!!!"
    />
    <div
      className="notes__image"
    >
      <img
        alt="imagen cheems"
        src="https://res.cloudinary.com/api-node-cloud/image/upload/v1644363562/journal-photos/r4rn9t0sczxtzhrbewfl.png"
      />
    </div>
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Eliminar
  </button>
</div>
`;
